source/loader/layers/sanitizer/asan_shadow.cpp

```diff
@@ -0,0 +1,232 @@
+        auto Result = getContext()->urDdiTable.VirtualMem.pfnFree(
+            Context, (const void *)ShadowBegin, GetShadowSize());
+        getContext()->urDdiTable.Context.pfnRelease(Context);
+        return Result;
+    }();
+    return Result;
+}
+
+ur_result_t ShadowMemoryGPU::EnqueuePoisonShadow(ur_queue_handle_t Queue,
+                                                 uptr Ptr, uptr Size,
+                                                 u8 Value) {
+    if (Size == 0) {
+        return UR_RESULT_SUCCESS;
+    }
+
+    uptr ShadowBegin = MemToShadow(Ptr);
+    uptr ShadowEnd = MemToShadow(Ptr + Size - 1);
+    assert(ShadowBegin <= ShadowEnd);
+    {
+        static const size_t PageSize =
+            GetVirtualMemGranularity(Context, Device);
+
+        ur_physical_mem_properties_t Desc{
+            UR_STRUCTURE_TYPE_PHYSICAL_MEM_PROPERTIES, nullptr, 0};
+
+        // Make sure [Ptr, Ptr + Size] is mapped to physical memory
+        for (auto MappedPtr = RoundDownTo(ShadowBegin, PageSize);
+             MappedPtr <= ShadowEnd; MappedPtr += PageSize) {
+            std::scoped_lock<ur_mutex> Guard(VirtualMemMapsMutex);
+            if (VirtualMemMaps.find(MappedPtr) == VirtualMemMaps.end()) {
+                ur_physical_mem_handle_t PhysicalMem{};
+                auto URes = getContext()->urDdiTable.PhysicalMem.pfnCreate(
+                    Context, Device, PageSize, &Desc, &PhysicalMem);
+                if (URes != UR_RESULT_SUCCESS) {
+                    getContext()->logger.error("urPhysicalMemCreate(): {}",
+                                               URes);
+                    return URes;
+                }
+
+                URes = getContext()->urDdiTable.VirtualMem.pfnMap(
+                    Context, (void *)MappedPtr, PageSize, PhysicalMem, 0,
+                    UR_VIRTUAL_MEM_ACCESS_FLAG_READ_WRITE);
+                if (URes != UR_RESULT_SUCCESS) {
+                    getContext()->logger.debug("urVirtualMemMap({}, {}): {}",
+                                               (void *)MappedPtr, PageSize,
+                                               URes);
+                    return URes;
+                }
+
+                getContext()->logger.debug("urVirtualMemMap: {} ~ {}",
+                                           (void *)MappedPtr,
+                                           (void *)(MappedPtr + PageSize - 1));
+
+                // Initialize to zero
+                URes = EnqueueUSMBlockingSet(Queue, (void *)MappedPtr, 0,
+                                             PageSize);
+                if (URes != UR_RESULT_SUCCESS) {
+                    getContext()->logger.error("EnqueueUSMBlockingSet(): {}",
+                                               URes);
+                    return URes;
+                }
+
+                VirtualMemMaps[MappedPtr].first = PhysicalMem;
+            }
+
+            // We don't need to record virtual memory map for null pointer.
```

Comment:
Why not record this?

Reply:
for null pointer, it doesn't have an alloc info.

Reply:
I also feel a little confused while reading this comment. 
I think we need to use nullpointer's shadow memory.

Reply:
Added more comments to describe the reasons.